### ISSUE TITLE
Problems with git seeing a diff if the model has images (zip binary format)

### ISSUE THREAD

**Issue by Phillipus** [archimatetool/archi-modelrepository-plugin2#2]
If a model contains images (and is therefore in a binary zip format) and you make a change to the model, save it, and then undo and save again git sees it as a changed file (SHA-1 is different).

Therefore we don't know if a model has really changed.

I think it is something  to do with the zip archiving library and compressing.

If the *.archimate file is in xml format at least git can detect changes for a quick check to see if a model needs committing.

Perhaps time to store images inline in the `*.archimate` file as base64?

**Comment by jbsarrodie** [comment]
Would it be possible to not use Zip format but simply its content as the working tree ? This should allow to reuse most of the current loading code.

**Comment by Phillipus** [comment]
> Would it be possible to not use Zip format but simply its content as the working tree ? This should allow to reuse most of the current loading code.

Good idea...but perhaps tricky.

We would then need two models when saving, as we have in coArchi 1 - the regular Archi saved model file, and the working tree version as well. Archi needs the binary `*.archimate` file to load.

Or we could change the Archi code to have an option to not save/load to zip format.

**Comment by Phillipus** [comment]
> Or we could change the Archi code to have an option to not save/load to zip format.

I've done this now in `com.archimatetool.editor.model.impl.ArchiveManager`.

Loading:

- If the model file is in archive format (zipped) then images are loaded from the zipped model file
- If the model file is _not_ in archive format and has image path references in the model then a relative `images/` directory is used to locate the image files

Saving:

- If `model.getAdapter("saveToArchiveFormat")` returns true then the archive format is used, else the model is saved to an `*.archimate` file and image files are saved to a relative `images/` directory

Now the only thing I need to figure out is when to set `model.setAdapter("saveToArchiveFormat", true)`.

**Comment by Phillipus** [comment]
> Now the only thing I need to figure out is when to set model.setAdapter("saveToArchiveFormat", true).

I think we may have to set this flag as a `Feature` in the model like:

`model.getFeatures().putBoolean("saveToArchiveFormat", false);`

And do this in these cases:

- The model is cloned from a repo (`CloneModelAction`) 
- A model is added to a repository (`CreateRepoFromModelAction`)

**Comment by Phillipus** [comment]
Need to think this through...

- It is technically possible to save a model with images in a "flat" format (an *.archimate file, and the image files in an `images` sub-folder)
- We only want to use this format for coArchi
- But this has to be done in the main Archi code (in `com.archimatetool.editor.model.impl.ArchiveManager`) because the user can save the model at any time, even if coArchi is not open
- We have to set a flag somewhere to denote that we are saving the model in this format
- If the flag is stored in the model as a `feature` then it is a problem if the user does a "Save As". Thereafter, the copy of the model will be saved in the flat format outside of coArchi because `ArchiveManager` sees the flag when saving

An alternative method, but inferior, would be one similar to coArchi 1 - save the model as normal to a `temp.archimate` file and then unzip this to the flat format in the working directory when committing.

**Comment by Phillipus** [comment]
> An alternative method, but inferior, would be one similar to coArchi 1 - save the model as normal to a temp.archimate file and then unzip this to the flat format in the working directory when committing. 

But this leaves us with the original problem:

> If a model contains images (and is therefore in a binary zip format) and you make a change to the model, save it, and then undo and save again git sees it as a changed file (SHA-1 is different).

coArchi 1 has this problem as well. If a model has images we save it (in zip format) to the `temp.archimate` file. Then, if we make a change, save it, then undo the change and save it again coArchi 1 will report that there are changes that need committing when there are not. This is because we store a checksum for the `temp.archimate` file to check for local changes.

**Comment by Phillipus** [comment]
I've thought of another way to do it that doesn't involve any changes to the Archi code, and is similar to the way we do it in coArchi 1.

- Save the actual model to `.git/temp.archimate` as we do in coArchi 1
- `ModelRepositoryPlugin` is a model listener for the model being saved (it already does this in coArchi 1). Once saved we copy the `temp.archimate` file to `model.archimate` in the working directory. If the model file is in zip format we copy its contents to the working directory so we end up with a `model.archimate` file and an `images` folder containing the image files. This allows us to do an instant check on whether something has changed and needs committing.

But there is a danger that the model might be saved when `ModelRepositoryPlugin` is not active (perhaps coArchi is uninstalled or for some other reason). So we can also perform the copy action just before important coArchi/git actions (Commit, Merge, Pull, etc...)

Basically we call `IArchiRepository#copyModelToWorkingDirectory` instead of calling `IArchiRepository#exportModelToGraficoFiles`

**Comment by Phillipus** [comment]
> I think it is something to do with the zip archiving library and compressing and the CRC is different each time.

It's because of the timestamp of each zip entry being different.

If we do this - `zipEntry.setTime(0);` - then the SHA-1 is the same each time we save the zip file.

However, setting the zip entries timestamp to the same each time might not be a good idea.

**Comment by jbsarrodie** [comment]
> It's because of the timestamp of each zip entry being different.
> 
> If we do this - `zipEntry.setTime(0);` - then the SHA-1 is the same each time we save the zip file (if nothing has changed).
> 
> However, setting the zip entries timestamp to the same each time might not be a good idea.

I think this is a good idea. We use a zip not to group and compress user files, but only to group information. I can't think of any good reason to set time on this zip content.

**Comment by Phillipus** [comment]
@jbsarrodie I've renamed this issue to reflect the real issue here.

I now have a method that works as detailed in the Comment above - https://github.com/archimatetool/archi-modelrepository-plugin2/issues/2#issuecomment-653890288

- It works and is fast
- It keeps the working dir files separate from the `model.archimate` file in case something goes wrong in a merge or other git action
- It avoids the problem of committing a binary file if it contains images
- If there is a simple change (name, documentation, new image, etc) we can easily see the change - again trying to avoid committing a binary file if it contains images

**Comment by jbsarrodie** [comment]
> I now have a method that works as detailed in the Comment above - [#2 (comment)](https://github.com/archimatetool/archi-modelrepository-plugin2/issues/2#issuecomment-653890288)

I'm not fully convinced this is needed...
 
>     * It works and is fast
Good

>     * It keeps the working dir files separate from the `model.archimate` file in case something goes wrong in a merge or other git action

Should not happen, unless of course if someone does a merge outside Archi, but this already leads to troubles with coArchi 1 and there's nothing we can do for that (and of course it will always be possible to restore a previous version to solve such issue)
 
>     * It avoids the problem of committing a binary file if it contains images

I don't see this as a problem. In fact, that's almost the opposite for me, as this at least make sure that git won't automagically try to merge the .archimate file by itself (which it absolutely must avoid).

>     * If there is a simple change (name, documentation, new image, etc) we can easily see the change - again trying to avoid committing a binary file if it contains images

I don't think this is of any help on a model file (even with coArchi 1 I rarely look at small XML files from git)

So for me all that we need is a simple and reliable way to know if a model has been edited. In this context, _reliable_ means that we can accept false positive (we think model has been changed but in fact was not), but not false negative (don't detect that a model changed).

**Comment by Phillipus** [comment]
OK, if we are going to do it that way then this is what we got:

Pros:

- The `model.archimate` file _is_ the model file
- No other temp model file is needed
- Even faster
- No tricks needed
- Makes it less confusing for users as they could in fact open _the_ model from desktop
- No extra code needed to copy/export/unpack etc
- git status can tell us when model changed / dirty

Cons (which may not matter but list them anyway):

- Binary models (with images) can not be viewed in git clients
- Will need to make changes to Archi code for things like zipping up the model with time set (see https://github.com/archimatetool/archi-modelrepository-plugin2/issues/2#issuecomment-665083737)
- Possible corruption of model file in merge conflict (but can restore from git)

> as this at least make sure that git won't automagically try to merge the .archimate file by itself (which it absolutely must avoid).

It could happen if the file is xml and not binary (no images).

**Comment by jbsarrodie** [comment]
My personal opinion on this...

> Cons (which may not matter but list them anyway):
> * Binary models (with images) can not be viewed in git clients

Users should not care about git clients. Power users would still be able to download the file and open it as any other model.

> * Will need to make changes to Archi code for things like zipping up the model with time set (see [#2 (comment)](https://github.com/archimatetool/archi-modelrepository-plugin2/issues/2#issuecomment-665083737))
> * Possible corruption of model file in merge conflict (but can restore from git)
> * Are there cases (troubleshooting) where user wants to see the raw xml file in git commits?

I do this sometime with coArchi 1, but only because XML files are really small and thus understandable. But even in this case, this is useful only with concepts, not views (always too big), so I can't think of any reasons to do this on a whole model.

> > as this at least make sure that git won't automagically try to merge the .archimate file by itself (which it absolutely must avoid).
> 
> It could happen if the file is xml and not binary (no images).

I know, I wanted to say, that if model file was _always_ binary instead of _always_ plain XML, this might be a good thing in this context.

**Comment by Phillipus** [comment]
> I know, I wanted to say, that if model file was always binary instead of always plain XML, this might be a good thing in this context.

I'm now making changes to the code to use one `model.archimate` file in the working directory (no more temp file, copying etc). I'd like to see how it behaves in this respect.

**Comment by Phillipus** [comment]
A `*.bak` file is created when a save is done. This makes it look like a change has been made. The only thing I can think to do is to create a `.gitgnore` file when creating/cloning a new repo.

**Comment by Phillipus** [comment]
@jbsarrodie If we are only ever going to commit one file, `model.archimate`, then we don't need gitignore - when we call git status or commit we can add a file pattern to only use this file.

**Comment by Phillipus** [comment]
I have this working now.

- The model is one file in the working directory - `model.archimate`
- All other files are ignored on commit and status
- For a binary file with images I had to make two changes in the main Archi code when saving model to zip format to set all file timestamps to zero (actually 01/01/1970 00:00 time) so that if you make a change and undo it git status will see an unchanged file

@jbsarrodie Before I do any more changes, is this how we wan things?

**Comment by jbsarrodie** [comment]
> @jbsarrodie Before I do any more changes, is this how we wan things?

I think so !

**Comment by Phillipus** [comment]
> I think so !

OK, in that case coArchi 2 will require the next version of Archi (because of the changes to zip timestamps when saving)

**Comment by jbsarrodie** [comment]
> OK, in that case coArchi 2 will require the next version of Archi (because of the changes to zip timestamps when saving)

Yes and no.

As there could be some "false positive" (file seen as changed while model content is not) at any time (e.g. slight variations in the XML serialization), we'll have to make sure nothing is done is we end up discovering that the model did not change. So this means, that coArchi 2 could still be used with current version of Archi, the only impact being on performance (it would take more time to see that there's nothing to commit).

**Comment by Phillipus** [comment]
> As there could be some "false positive" (file seen as changed while model content is not) at any time (e.g. slight variations in the XML serialization), we'll have to make sure nothing is done is we end up discovering that the model did not change. So this means, that coArchi 2 could still be used with current version of Archi, the only impact being on performance (it would take more time to see that there's nothing to commit).

I wonder if XML serialization does vary even if the model content is the same?

Comparing two logical models to see if there was an actual change might be time consuming. Also this would be non-trivial.

**Comment by jbsarrodie** [comment]
> Also this would be non-trivial.

But this will be needed anyway for merging, so we reuse the code here.

> Comparing two logical models to see if there was an actual change might be time consuming.

Yes, but as there is a need to know what has been changed in the model, we have to compare previous and current version anyway...

**Comment by Phillipus** [comment]
> I wonder if XML serialization does vary even if the model content is the same?

Yes, it does. Simply moving some concepts from one folder to another and then back again (not undo) can change the order of the elements in the XML file and thus will appear to be a change.

So I'm not sure if there's any point in setting the timestamps to zero in a zip format file.

**Comment by jbsarrodie** [comment]
>So I'm not sure if there's any point in setting the timestamps to zero in a zip format file.

I think there are some (but few) cases where it makes sens, so this is a cost vs benefit choice: benefit is low, so if the code change is simple and straightforward to set timestamps to zero, then do it, if this is more complicated, don't do it.

**Comment by Phillipus** [comment]
This is solved in the latest work done on coArchi 2 and in the main Archi code in `save-images-folder` branch.

**Comment by jbsarrodie** [comment]
Note to self:

> * If the flag is stored in the model as a `feature` then it is a problem if the user does a "Save As" (or "Save As Template" or "Save Canvas As Template"). Thereafter, the copy of the model will be saved in the flat format outside of coArchi because `ArchiveManager` sees the flag when saving

In fact this could be easily solved by forcing any "Save as" action to use the usual zip saving method before saving.

**Comment by Phillipus** [comment]
Model is saved as `model.archimate` file and `/images` folder in the working dir. Archi can load it and save it in that format directly without any tricks, this is in main Archi code in `save-images-folder` branch.

We mustn't save it in the zip format because the repo will get too massive - 10mb file with one small change is another 10mb file added to the git repo, and another, etc, until the repo size is 100s of mb. I tested that.
